zng/cc.rs

```diff
@@ -0,0 +1,470 @@
+use std::{
+    env, fs,
+    io::Write as _,
+    path::{Path, PathBuf},
+};
+
+struct Build {
+    cfg: cc::Build,
+    is_msvc: bool,
+}
+
+impl Build {
+    fn new(cfg: cc::Build) -> Self {
+        let is_msvc = cfg.get_compiler().is_like_msvc();
+        Self { cfg, is_msvc }
+    }
+
+    fn append(&mut self, root: Option<&str>, files: &[&str]) {
+        let root = root.map_or(String::new(), |s| {
+            assert!(!s.ends_with('/'), "remove trailing slash");
+            format!("{s}/")
+        });
+        self.cfg.files(
+            files
+                .into_iter()
+                .map(|fname| format!("src/zlib-ng/{root}{fname}.c")),
+        );
+    }
+
+    fn mflag(
+        &mut self,
+        non_msvc: impl Into<Option<&'static str>>,
+        msvc: impl Into<Option<&'static str>>,
+    ) {
+        let Some(flag) = (if self.is_msvc {
+            msvc.into()
+        } else {
+            non_msvc.into()
+        }) else {
+            return;
+        };
+        self.cfg.flag(flag);
+    }
+}
+
+impl std::ops::Deref for Build {
+    type Target = cc::Build;
+
+    fn deref(&self) -> &Self::Target {
+        &self.cfg
+    }
+}
+
+impl std::ops::DerefMut for Build {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.cfg
+    }
+}
+
+/// Replicate the behavior of cmake/make/configure of stripping out the
+/// @ZLIB_SYMBOL_PREFIX@ since we don't want or need it
+fn strip_symbol_prefix(input: &Path, output: &Path, get_version: bool) -> String {
+    let contents = fs::read_to_string(input)
+        .map_err(|err| format!("failed to read {input:?}: {err}"))
+        .unwrap();
+    let mut h =
+        std::io::BufWriter::new(fs::File::create(output).expect("failed to create zlib include"));
+
+    use std::io::IoSlice;
+    let mut write = |bufs: &[IoSlice]| {
+        // write_all_vectored is unstable
+        for buf in bufs {
+            h.write_all(&buf).unwrap();
+        }
+    };
+
+    let mut version = None;
+    for line in contents.lines() {
+        if let Some((begin, end)) = line.split_once("@ZLIB_SYMBOL_PREFIX@") {
+            write(&[
+                IoSlice::new(begin.as_bytes()),
+                IoSlice::new(end.as_bytes()),
+                IoSlice::new(b"\n"),
+            ]);
+        } else {
+            write(&[IoSlice::new(line.as_bytes()), IoSlice::new(b"\n")]);
+        }
+
+        if get_version {
```

Comment:
To avoid the `contains` call after you find the version.

```suggestion
        if version.is_none() && get_version {
```